Infrastructure as code (#111)
RasaHQ/financial-demo: README.md

You will need to have docker installed in order to build the action server image. If you haven't made any changes to the action code, you can also use the [public image on Dockerhub](https://hub.docker.com/r/rasa/financial-demo) instead of building it yourself.

> Build & tag the image:

Reviewer: ```suggestion
To build & tag the image, run:
```

```bash
export ACTION_SERVER_DOCKERPATH=<dockerID>/<name-of-image>:<tag-of-image>
make docker-build
```

Run the action server container:

```bash
make docker-run
```

Perform a smoke test on the health endpoint:

```bash
make docker-test
```

> Once you have confirmed that the container works as it should, push the container image to a registry:

Reviewer: ```suggestion
Once you have confirmed that the container is working, push the container image to a registry:
```

```bash
# login to a container registry with your credentials
docker login  

# check the registry logged into
docker system info | grep Registry

# push the action server image
make docker-push
```

> # CI/CD

Reviewer: Can you add the CI/CD section to the README's table of contents?

Author: Done!

## Summary

> A CI/CD pipeline is used to test, build and deploy the financial-demo bot to AWS EKS:

Reviewer: I think it'd make sense to link the docs page here: https://rasa.com/docs/rasa/setting-up-ci-cd

Author: Done!

The pipeline uses GitHub Actions, defined in  `.github/workflows/cicd.yml`. It includes these jobs:

![](images/cicd.png)

> **params**

Reviewer: Can you link to the specific lines in cicd.yml per section here?

Author: That would be really nice indeed. How would that be done though?
One way would be to get a permalink to the lines of the cicd.yml in the GitHub repo, but I wonder if that is robust. What happens when the cicd.yml is being updated?

Reviewer: A permalink should work in the case that the file gets updated: https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

Author: Done!

- Defines parameters for use by downstream jobs

**params_summary**

- Prints the value of the parameters.

**action_server**

- Builds & Tests the docker image of the action server with tag: `<branch-name>`
- Uploads the docker image to the AWS ECR repository: `financial-demo`

**rasa_model**

- Trains & Tests the rasa model with name: `models/<branch-name>.tar.gz
- Uploads the trained model to the AWS S3 bucket: `rasa-financial-demo` 

**aws_eks_create_test_cluster**

> - If not existing yet, create an AWS EKS cluster with name: `financial-demo-<branch-name>`

Reviewer: ```suggestion
- If not existing yet, creates an AWS EKS cluster with name: `financial-demo-<branch-name>`
```

**deploy_to_test_cluster**

- Installs/Updates Rasa Enterprise, with the docker image created by the **action_server** job.
- Deploys the rasa model, trained by the **rasa_model** job.

> - Performs smoke tests to ensure basic operation is all OK.

Reviewer: ```suggestion
- Performs smoke tests to ensure basic operations are all OK.
```

Reviewer: Could you add a section documenting what the smoke tests are doing as well?

**deploy_to_prod_cluster**

- Runs when pushing to the `main` branch, and all previous steps are successful.
- Installs/Updates Rasa Enterprise, with the docker image created by the **action_server** job.
- Deploys the rasa model, trained by the **rasa_model** job.

> - Performs smoke tests to ensure basic operation is all OK.

Reviewer: ```suggestion
- Performs smoke tests to ensure basic operations are all OK.
```

## GitHub Secrets

> In your GitHub repository, go to `Settings > Secrets`, and add these `New repository secrets` . 

Reviewer: ```suggestion
Secrets can be added to your GitHub repository by going to `Settings > Secrets` and selecting `New repository secret`. 
```

> When entering the value, do not use quotes.

Reviewer: ```suggestion
When entering values, be sure to omit quotes.
```

#### AWS IAM User API Keys:

> To allow the GitHub actions to configure the aws cli, create IAM User API Keys as described below, and add them as GitHub Secrets to the repo:

Reviewer: ```suggestion
To configure the aws cli in Github Actions, create IAM User API Keys as described below, and add them as GitHub secrets to the repository:
```

Reviewer: Can you add a link to the AWS CLI docs here as well?

- AWS_ACCESS_KEY_ID = `Access key ID`
- AWS_SECRET_ACCESS_KEY = `Secret access key` 

#### AWS Elastic IP:

Create an Elastic IP as described below, and add it as a GitHub Secret to the repo:

- AWS_ELASTIC_IP = `PublicIp`

#### Rasa Enterprise License:

> To allow the GitHub actions to define a pull secret for the private GCR repo, get the private values from your Rasa Enterprise license file ([docs](https://rasa.com/docs/rasa-x/installation-and-setup/install/helm-chart#5-configure-rasa-x-image)) and add them as GitHub Secrets to the repo:

Reviewer: ```suggestion
To define a pull secret in Github Actions for the private GCR repo, you'll need to retrieve the private values from your Rasa Enterprise license file ([docs](https://rasa.com/docs/rasa-x/installation-and-setup/install/helm-chart#5-configure-rasa-x-image)) and add them as GitHub secrets to the repository:
```

- GCR_AUTH_JSON_PRIVATE_KEY_ID = `private_key_id`
- GCR_AUTH_JSON_PRIVATE_KEY = `private_key`
- GCR_AUTH_JSON_CLIENT_EMAIL = `client_email`
- GCR_AUTH_JSON_CLIENT_ID = `client_id`

#### Helm chart Credentials

> To allow the GitHub actions to use safe_credentials in the `values.yml` ([docs](https://rasa.com/docs/rasa-x/installation-and-setup/install/helm-chart#3-configure-credentials)), add following GitHub Secrets to the repo, replacing each `<safe credential>` with a different alphanumeric string, and choose a `<username>` for the initial user. 

Reviewer: ```suggestion
To use safe_credentials in Github Actions through `values.yml` ([docs](https://rasa.com/docs/rasa-x/installation-and-setup/install/helm-chart#3-configure-credentials)), add following GitHub Secrets to the repo, replacing each `<safe credential>` with a different alphanumeric string and choosing a `<username>` for the initial user. 
```

*(Please use **safe credentials** to avoid data breaches)*

- GLOBAL_POSTGRESQL_POSTGRESQLPASSWORD = `<safe credential>`
- GLOBAL_REDIS_PASSWORD = `<safe credential>`
- RABBITMQ_RABBITMQ_PASSWORD = `<safe credential>`
- RASAX_INITIALUSER_USERNAME = `<username>`
- RASAX_INITIALUSER_PASSWORD = `<safe credential>`
- RASAX_JWTSECRET = `<safe credential>`
- RASAX_PASSWORDSALT = `<safe credential>`
- GLOBAL_POSTGRESQL_POSTGRESQLPASSWORD = `<safe credential>`
- GLOBAL_POSTGRESQL_POSTGRESQLPASSWORD = `<safe credential>`

## AWS Preparation

The CI/CD pipeline of financial-demo uses AWS for all the storage & compute resources.

After cloning or forking the financial-demo GitHub repository you must set up the following items before the pipeline can run.

### IAM User API Keys

The CI/CD pipeline uses the [aws cli](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html#cliv2-linux-install).

The [aws cli](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html#cliv2-linux-install) needs a set of IAM User API keys for authentication & authorization:

> - In your AWS Console, go to IAM dashboard to create a new set of API keys:

Reviewer: ```suggestion
In your AWS Console, go to the IAM dashboard to create a new set of API keys:
```

  - Click on Users

  - Click on Add user

    - User name = findemo  *(The actual name is not important, we will never use this name directly)*

      Choose "**programmatic access**." This allows you to use the aws cli to interact with AWS.

    - Click on Next: Permissions

      - Click on *Attach existing policies directly*

>         For IAM access, you can choose “**AdministratorAccess**”, or limit access to only what is needed by the CD pipeline.

Reviewer: Do you know what specific permissions would be required in AWS here?

Author: That will require some trial & error I think.
What if we create an issue for this and table that investigation for after the merge.
I agree it would be very helpful to have that information.l

Author: I created issue #121

    - Click on Next: Tags

    - Click on Next: Review

    - Click on Create user

    - Store in a safe location: `Access key ID` & `Secret access key`

### SSH Key Pair

> To be able to SSH into the EC2 worker nodes of the EKS cluster, you need an SSH Key Pair

Reviewer: ```suggestion
To SSH into the EC2 worker nodes of the EKS cluster, you'll need an SSH Key Pair
```

- In your AWS Console, go to **EC2 > Key Pairs**, and create a Key Pair with the name `findemo`, and download the file `findemo.pem` which contains the private SSH key. 
  **Note that the name `findemo` is important, since it is used by the CI/CD pipeline when the cluster is created.**

### Local AWS CLI

> Before the CI/CD pipeline can run, you will use the AWS CLI locally to create some items, so also install & configure the CLI locally.

Reviewer: ```suggestion
Before the CI/CD pipeline can run, you will use the AWS CLI locally to create some resources and need to install & configure the CLI locally.
```

AWS Secret Access Key [None]: -------    # See above: IAM User API Keys
Default region name [None]: us-west-2    # The CI/CD pipeline uses us-west-2 
Default output format [None]: 

# Check your configuration
aws configure list [--profile profile-name]

# verify it works
aws s3 ls
```

### ECR repository & S3 bucket

The CI pipeline creates two artifacts:

- An action server docker **image**, which is pushed to an **AWS ECR** repository.
- A trained rasa **model**, which is copied to an **AWS S3** bucket

Run these commands to create the storage for these artifacts. These commands run the AWS CLI and create an ECR repository and an S3 bucket:

```bash
# create ECR repository with name `financial-demo`
make aws-ecr-create-repository

# create S3 bucket with name `rasa-financial-demo`
make aws-s3-create-bucket
```

## EKS production cluster

If the production cluster is already set up, you can skip this section.

This section describes the initial deployment of the financial-demo bot on an EKS production cluster. 

This initial deployment is done manually. After that, the deployment is maintained & upgraded automatically by the CI/CD pipeline.

> ### Preparation

Reviewer: Can you add a brief sentence or two here summarizing what's needed for the preparation step?

Author: The **Preparation** section starts on the next line, so I added a summary there. 
Let me know if that works.

#### Install eksctl

See the [installation instructions](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html)

> If you use Ubuntu, you can just issue the command:

Reviewer: ```suggestion
If you use Ubuntu, you can issue the command:
```

```bash
make install-eksctl
```

#### Install kubectl

See the [installation instructions](https://kubernetes.io/docs/tasks/tools/#kubectl)

> If you use Ubuntu, you can just issue the command:

Reviewer: ```suggestion
If you use Ubuntu, you can issue the command:
```

```bash
make install-kubectl
```

#### Install helm

See the [installation instructions](https://helm.sh/docs/intro/install/)

> If you use Ubuntu, you can just issue the command:

Reviewer: ```suggestion
If you use Ubuntu, you can issue the command:
```

```bash
make install-helm
```

#### Install jp

See the [installation instructions](https://github.com/jmespath/jp#installing)

> If you use Ubuntu, you can just issue the command:

Reviewer: ```suggestion
If you use Ubuntu, you can issue the command:
```

```bash
make install-jp
```

#### Set environment variables

There are many ways to set the required environment variables in your local environment. 

> One way is to create a file  `./secret/envs_export`, with this content:

Reviewer: Can the GCR keys be set using the GCP CLI? Might be worth mentioning here as an alternative option.

Author: Wouldn't the GCP CLI be used only when you do things on a GCP VM?

This is just for defining the environments on your local computer.

Reviewer: They can authenticate into the registry using the CLI as well: https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud-helper. I think a link to their docs would be good enough if anyone is looking to authenticate a different way

Author: I added a link in the README.

```bash
# source this file to set the environment variables the same as the GitHub secrets.
export GCR_AUTH_JSON_PRIVATE_KEY_ID=...
export GCR_AUTH_JSON_PRIVATE_KEY='-----BEGIN PRIVATE KEY-...-END PRIVATE KEY-----\n'
export GCR_AUTH_JSON_CLIENT_EMAIL='...'
export GCR_AUTH_JSON_CLIENT_ID=...
export GLOBAL_POSTGRESQL_POSTGRESQLPASSWORD=...
export GLOBAL_REDIS_PASSWORD=...
export RABBITMQ_RABBITMQ_PASSWORD=...
export RASAX_INITIALUSER_USERNAME=admin
export RASAX_INITIALUSER_PASSWORD=...
export RASAX_JWTSECRET=...
export RASAX_PASSWORDSALT=...
export RASAX_TOKEN=...
export RASA_TOKEN=...
export AWS_ELASTIC_IP=...
```

Then, create the environment variables with the command:

```bash
source ./secret/envs_export
```

### Create the EKS cluster

Create the EKS production cluster, with name `financial-demo-<current branch name>`:

```bash
make aws-eks-cluster-create AWS_EKS_CLUSTER_NAME=financial-demo-production
```

> Some other useful 'Makefile functions' to interact with the EKS cluster:

Reviewer: I think it'd be helpful to create a separate section in the README documenting all the makefile functions and what they do.

Author: I was thinking about it, but I didn't do it because it introduces something that might be hard to maintain.
I also wonder how often someone would  read these explanation in the README, instead of just relying on the name of the makefile functions. I tried to make them kind of self explanatory.

Author: @b-quachtran 
Ah, yes, thank you!
I had removed the help section while developing the Makefile but forgot to put it back in.

Please check how it looks.

Install/Upgrade Rasa Enterprise, using the action server docker image that was uploaded to the ECR repository:

```bash
# make sure kubectl is looking at the correct EKS cluster
# -> if not, configure kubeconfig as described above
make kubectl-config-current-context  AWS_EKS_CLUSTER_NAME=financial-demo-production

# create namespace `my-namespace`
make aws-eks-namespace-create

# create/refresh gcr-pull-secret, for Rasa Enterprise image
make pull-secret-gcr-create

# create/refresh ecr-pull-secret, for action server image
make pull-secret-ecr-create

# Verify that:
#
# the correct Rasa Enterprise version is selected in `Makefile`
# ==> RASAX_TAG := ......
# 
# the correct rasa version is selected in `requirements.txt`
# ==> rasa[spacy]==....
#
# Install/Upgrade Rasa Enterprise with the action server
make rasa-enterprise-install

# Check Rasa Enterprise Health
make rasa-enterprise-check-health

# To troubleshoot, highly recommended to use OCTANT, see Appendix C
```

### Train, test & upload model to S3

```bash
# Train the model: `models/<current branch>.tar.gz`
make rasa-train

# In another window, start duckling server
docker run -p 8000:8000 rasa/duckling

> # Run the end-2-end tests

Reviewer: ```suggestion
# Run the end-to-end tests
```

make rasa-test

# Upload `models/<current branch>.tar.gz` to S3
# Note: This does not mean the model is deployed to Rasa Enterprise,
#       which is done in the next step.
make aws-s3-upload-rasa-model
```

### Deploy, Tag & Smoketest the trained model

```bash
# Configure ~/.kube/config and set current_context
make aws-eks-cluster-update-kubeconfig  AWS_EKS_CLUSTER_NAME=financial-demo-production

# Deploy rasa model
make aws-s3-download-rasa-model
make rasa-enterprise-model-delete
make rasa-enterprise-model-upload
make rasa-enterprise-model-tag

# Wait about 1 minute, so rasa-production can download, upack & load the model
# Smoketest
make rasa-enterprise-smoketest
```

> ### DNS

Reviewer: This step is optional right?

Author: Correct. 
I explained that it is optional and what it is for.

aws-financial-demo.my-domain.com canonical name = ---.us-west-2.elb.amazonaws.com.

########################################################################
# use `dig`
dig CNAME aws-financial-demo.my-domain.com
...
;; ANSWER SECTION:
aws-financial-demo.my-domain.com. ---- IN CNAME ---.us-west-2.elb.amazonaws.com.

```

Once propagated, you can access Rasa Enterprise at **http://aws-financial-demo.my-domain.com**

## Appendix A: The AWS EKS cluster

We use *eksctl* to create the clusters. It creates many AWS resources with CloudFormation templates. 

Some of these AWS resources are:

- A VPC with
  - Public & private subnets
  - Internet & NAT Gateways
  - Route Tables
- An IAM EKS service role
- An EKS Cluster Control Plane
- An EKS Managed nodegroup of EC2 instances

The cluster context is also added to the `~/.kube/config` file.

The VPC created looks something like this:

![](images/financial-demo-eks-test-vpc.png)

The EKS Control Plane interacts with the the EKS Data Plane (the nodes), like this:

![img](https://d2908q01vomqb2.cloudfront.net/fe2ef495a1152561572949784c16bf23abb28057/2020/04/10/subnet_pubpri.png)

## Appendix B: Manual Cleanup of AWS resources

> Sometimes things do not clean up properly, and you need to do a manual cleanup in the **AWS console**:

Reviewer: ```suggestion
Sometimes things do not clean up properly and you will need to do a manual cleanup in the **AWS console**:
```

> - **CloudFormation**: Try to delete all the stacks in reverse order as they were created by the eksctl command.

Reviewer: ```suggestion
**CloudFormation**: Try to delete all the stacks in reverse order as they were created by the eksctl command.
```

> - When a stack fails to delete due to dependencies, you have two options:

Reviewer: ```suggestion
When a stack fails to delete due to dependencies, you have two options:
```

  - Select to retain the resources that have dependency errors. (**NOT RECOMMENDED**)

    The stack delete operation will simply skip deleting them. This is NOT recommended, because you will clutter up your AWS account with many unused resources.

  - Manually delete the resources that the stack is not able to delete. (**RECOMMENDED**)

>     You do this by drilling down into the **CloudFormation stack delete events** messages, and delete items bottom-up the dependency tree.

Reviewer: ```suggestion
    You can do this by drilling down into the **CloudFormation stack delete events** messages and deleting items bottom-up the dependency tree.
```

    One example of a bottom-up delete sequence is when deletion of the VPC fails:

    - **EC2 > Load Balancers**: first, delete the ELB load balancers
    - **VPC > Subnets**: then, delete the subnets 
      - This will also delete the EC2 > Network interfaces, named `eni-xxxx`
      - You cannot delete Subnets until the ELB load balancers are deleted
    - **VPC > Your VPCs**: finally, delete the VPC 
      - This will also delete all associated:
        - security groups (`sg-xxx`)
        - internet gateways (`igw-xxx`)
        - subnets (`subnet-xxx`)

  - After cleaning up, try again to delete the **AWS CloudFormation** stack.

  - If it still does not delete, iterate the manual cleanups until it does.

    Again, this can be a painful process, but once the CloudFormation stacks delete properly without retaining/skipping anything, you are guaranteed that you have cleaned up all the resources created by the CI/CD pipeline.

## Appendix C: OCTANT

> [Octant](https://octant.dev/) is a great tool to look  inside the cluster and trouble shoot issues when they arise.

Reviewer: ```suggestion
[Octant](https://octant.dev/) is a useful open sourced tool for visualizing workloads inside the cluster and troubleshooting issues when they arise.
```

# #################################################### #

# -> for a feature branch test cluster
git checkout my-feature-branch

# -> for the production cluster
git branch production # create temporary, local `production` branch
git checkout production

# Configure kubeconfig for the correct EKS cluster
make aws-eks-cluster-update-kubeconfig

# make sure kubectl is now looking at the correct EKS cluster
make kubectl-config-current-context

# #################################################### #
# Run Octant on default port 7777 and open the browser #
# #################################################### #
OCTANT_LISTENER_ADDR=0.0.0.0:7777 octant &

# Within octant, select the namespace `my-namespace`
```

## Appendix D: AWS EKS references

There are two commands to create AWS EKS clusters;  `eksctl` & `aws eks`.  

- The `eksctl` cli is the most complete, and it is what we ended up using.

- The `aws eks` cli does not support to launch worker nodes to the cluster control plane. This has to be done manually from the AWS Console, which makes it unsuited for a CI/CD pipeline where everything needs to be done via scripting (=> infrastructure as code).

> The following references are great to learn about AWS EKS, and it is highly recommended to manually build some test-clusters with both the `eksctl` & `aws eks` commands to demystify all those AWS resources that are being generated:

Reviewer: ```suggestion
The following references are useful for learning about AWS EKS, and it is highly recommended to manually build test-clusters with both the `eksctl` & `aws eks` commands to help demystify the AWS resources that are being generated:
```
